src/servers/src/mysql/handler.rs

```diff
@@ -140,34 +165,59 @@ impl<W: AsyncWrite + Send + Sync + Unpin> AsyncMysqlShim<W> for MysqlInstanceShi
         true
     }
 
-    async fn on_prepare<'a>(&'a mut self, _: &'a str, w: StatementMetaWriter<'a, W>) -> Result<()> {
-        w.error(
-            ErrorKind::ER_UNKNOWN_ERROR,
-            b"prepare statement is not supported yet",
-        )
-        .await?;
-        Ok(())
+    async fn on_prepare<'a>(
+        &'a mut self,
+        query: &'a str,
+        w: StatementMetaWriter<'a, W>,
+    ) -> Result<()> {
+        let (query, param_num) = replace_placeholder(query);
+        if let Err(e) = validate_query(&query).await {
+            w.error(ErrorKind::ER_UNKNOWN_ERROR, e.to_string().as_bytes())
+                .await?;
+            return Ok(());
+        };
+
+        let stmt_id = self.set_query(query);
+        let params = dummy_params(param_num);
+
+        w.reply(stmt_id, &params, &[]).await?;
+        return Ok(());
     }
 
     async fn on_execute<'a>(
         &'a mut self,
-        _: u32,
-        _: ParamParser<'a>,
+        stmt_id: u32,
+        p: ParamParser<'a>,
         w: QueryResultWriter<'a, W>,
     ) -> Result<()> {
-        w.error(
-            ErrorKind::ER_UNKNOWN_ERROR,
-            b"prepare statement is not supported yet",
-        )
-        .await?;
+        let params: Vec<ParamValue> = p.into_iter().collect();
+        let query = match self.query(stmt_id) {
+            None => {
+                w.error(
+                    ErrorKind::ER_UNKNOWN_STMT_HANDLER,
+                    b"prepare statement not exist",
+                )
+                .await?;
+                return Ok(());
+            }
+            Some(query) => query,
+        };
+
+        let query = replace_params(params, query);
+        log::debug!("execute replaced query: {}", query);
+
+        let outputs = self.do_query(&query).await;
+        write_output(w, &query, outputs).await?;
+
         Ok(())
     }
 
-    async fn on_close<'a>(&'a mut self, _stmt_id: u32)
+    async fn on_close<'a>(&'a mut self, stmt_id: u32)
     where
         W: 'async_trait,
     {
-        // do nothing because we haven't implemented prepare statement
+        let mut guard = self.prepared_stmts.write();
+        guard.remove(&stmt_id);
```

Comment:
Can we use some crate like [moka](https://crates.io/crates/moka) to cache prepared statements instead of plain hashmap? It's hard to achieve TTL in plain hashmap, which is a require with caching.

Reply:
You are right, this implementation is just a baby step, I'll continue to work on the datafusion type infer approach.
How about add a TODO here?

Reply:
Sure.